stumpy/mmotifs.py

```diff
@@ -0,0 +1,210 @@
+        (a candidate motif) and all subsequences in T to be considered as a match.
+        If None, this defaults to
+        `np.nanmax([np.nanmean(D) - 2 * np.nanstd(D), np.nanmin(D)])`
+        (i.e. at least the closest match will be returned).
+
+    cutoffs: numpy.ndarray or float, default None
+        The largest matrix profile value (distance) for each dimension of the
+        multidimensional matrix profile that a multidimenisonal candidate motif is
+        allowed to have.
+        If cutoffs is only one value, these value will be applied to every dimension.
+
+    max_matches: int, default 10
+        The maximum amount of similar matches (nearest neighbors) of a motif
+        representative to be returned.
+        The first match is always the self-match for each motif.
+
+    max_motifs: int, default 1
+        The maximum number of motifs to return
+
+    atol : float, default 1e-8
+        The absolute tolerance parameter. This value will be added to `max_distance`
+        when comparing distances between subsequences.
+
+    normalize : bool, default True
+        When set to `True`, this z-normalizes subsequences prior to computing distances.
+        Otherwise, this function gets re-routed to its complementary non-normalized
+        equivalent set in the `@core.non_normalized` function decorator.
+
+    p : float, default 2.0
+        The p-norm to apply for computing the Minkowski distance. This parameter is
+        ignored when `normalize == False`.
+
+    Returns
+    -------
+    motif_distances: numpy.ndarray
+        The distances corresponding to a set of subsequence matches for each motif.
+
+    motif_indices: numpy.ndarray
+        The indices corresponding to a set of subsequences matches for each motif.
+
+    motif_subspaces: numpy.ndarray
+        A numpy.ndarray consisting of arrays that contain the `k`-dimensional
+        subspace for each motif.
+
+    motif_mdls: numpy.ndarray
+        A numpy.ndarray consisting of arrays that contain the mdl results for
+        finding the dimension of each motif
+    """
+    # Convert dataframe to array if necessary
+    T = core._preprocess(T)
+
+    if max_motifs < 1:  # pragma: no cover
+        logger.warning(
+            "The maximum number of motifs, `max_motifs`, "
+            "must be greater than or equal to 1"
+        )
+        logger.warning("`max_motifs` has been set to `1`")
+        max_motifs = 1
+
+    # Calculate subsequence length
+    m = T.shape[-1] - P.shape[-1] + 1
+
+    # Calculate exclusion zone
+    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
+
+    # Set default max_matches value
+    if max_matches is None:
+        max_matches = np.inf
+
+    # Calculate default cutoff values
+    if cutoffs is None:
+        cutoffs = []
+        P_copy = P.copy().astype(np.float64)
+        P_copy[np.isinf(P_copy)] = np.nan
+        for i in range(P_copy.shape[0]):

```

Comment:
Instead of using a for-loop and appending to a list, I think we can leverage the `axis` parameter and do:

```
cutoffs = np.nanmax(
    [
        np.nanmean(P_copy - 2.0 * np.nanstd(P_copy, axis=0), axis=0),
        np.nanmin(P_copy, axis=0),
    ],
    axis=0,
)
```

So, `cutoffs` can be either a list OR a 1-D NumPy array. In both cases, indexing would be the same

Reply:
I found out that it is only working for 
`cutoffs = np.nanmax(
    [
        np.nanmean(P_copy - 2.0 * np.nanstd(P_copy, axis=0), axis=1),
        np.nanmin(P_copy, axis=1),
    ],
    axis=0,
)`

However, I am completely confused right now since I get difficult values for the first entry of the array when using the previous computation and the current (new) calculation. But all other entries are completely identical.
I'm probably just making a totally stupid mistake, but I get the following results for `cutoffs`:
Previous:
![grafik](https://user-images.githubusercontent.com/95933072/154255522-f7396f7d-ce28-47fa-9c53-3be70549a7ff.png)
New:
![grafik](https://user-images.githubusercontent.com/95933072/154255571-d711eae9-0355-4cdb-bb93-62a579eebdb0.png)

Reply:
Please see this response: https://github.com/TDAmeritrade/stumpy/pull/544#discussion_r808023666